Code/Framework/AzCore/AzCore/Memory/AllocatorBase.cpp

```diff
@@ -11,10 +11,157 @@
-    IAllocator(allocationSource),
-    m_name(name),
-    m_desc(desc)
+#define RECORDING_ENABLED 0
+
+#if RECORDING_ENABLED
+
+#include <AzCore/std/containers/map.h>
+#include <AzCore/IO/SystemFile.h>
+#include <AzCore/std/parallel/mutex.h>
+#include <AzCore/std/parallel/scoped_lock.h>
+
+namespace
+{
+    class DebugAllocator
+    {
+    public:
+        typedef void* pointer_type;
+        typedef AZStd::size_t size_type;
+        typedef AZStd::ptrdiff_t difference_type;
+        typedef AZStd::false_type allow_memory_leaks; ///< Regular allocators should not leak.
+
+        AZ_FORCE_INLINE pointer_type allocate(size_t byteSize, size_t alignment, int = 0)
+        {
+            return AZ_OS_MALLOC(byteSize, alignment);
+        }
+        AZ_FORCE_INLINE size_type resize(pointer_type, size_type)
+        {
+            return 0;
+        }
+        AZ_FORCE_INLINE void deallocate(pointer_type ptr, size_type, size_type)
+        {
+            AZ_OS_FREE(ptr);
+        }
+    };
+
+    #pragma pack(push, 1)
+    struct alignas(1) AllocatorOperation
+    {
+        enum OperationType : size_t
+        {
+            ALLOCATE,
+            DEALLOCATE
+        };
+        OperationType m_type: 1;
+        size_t m_size : 28; // Can represent up to 256Mb requests
+        size_t m_alignment : 7; // Can represent up to 128 alignment
+        size_t m_recordId : 28; // Can represent up to 256M simultaneous requests, we reuse ids
+    };
+    #pragma pack(pop)
+    static_assert(sizeof(AllocatorOperation) == 8);
+
+    static AZStd::mutex s_operationsMutex = {};
+
+    static constexpr size_t s_maxNumberOfAllocationsToRecord = 16384;
+    static size_t s_numberOfAllocationsRecorded = 0;
+    static constexpr size_t s_allocationOperationCount = 5 * 1024;
+    static AZStd::array<AllocatorOperation, s_allocationOperationCount> s_operations = {};
+    static uint64_t s_operationCounter = 0;
+
+    static unsigned int s_nextRecordId = 1;
+    using AllocatorOperationByAddress = AZStd::map<void*, AllocatorOperation, AZStd::less<void*>, DebugAllocator>;
+    static AllocatorOperationByAddress s_allocatorOperationByAddress;
+    using AvailableRecordIds = AZStd::vector<unsigned int, DebugAllocator>;
+    AvailableRecordIds s_availableRecordIds;
+
+    void RecordAllocatorOperation(AllocatorOperation::OperationType type, void* ptr, size_t size = 0, size_t alignment = 0)
+    {
+        AZStd::scoped_lock<AZStd::mutex> lock(s_operationsMutex);
+        if (s_operationCounter == s_allocationOperationCount)
+        {
+            AZ::IO::SystemFile file;
+            int mode = AZ::IO::SystemFile::OpenMode::SF_OPEN_APPEND | AZ::IO::SystemFile::OpenMode::SF_OPEN_WRITE_ONLY;
+            if (!file.Exists("memoryrecordings.bin"))
+            {
+                mode |= AZ::IO::SystemFile::OpenMode::SF_OPEN_CREATE;
+            }
+            file.Open("memoryrecordings.bin", mode);
+            if (file.IsOpen())
```

Comment:
Can you add a comment indicating that `memoryrecordings.bin` is being output to the current working directory?